src/HttpClient/HttpPluginClientBuilder.php

```diff
@@ -12,26 +12,51 @@
 use Http\Client\Common\HttpMethodsClient;
 use Http\Client\Common\Plugin;
 use Http\Client\Common\PluginClient;
-use Http\Discovery\HttpClientDiscovery;
-use Http\Discovery\MessageFactoryDiscovery;
+use Http\Discovery\Psr17FactoryDiscovery;
+use Http\Discovery\Psr18ClientDiscovery;
 use Http\Message\RequestFactory;
 use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
 
 class HttpPluginClientBuilder
 {
     /** @var ClientInterface */
     private $httpClient;
     /** @var HttpMethodsClient|null */
     private $pluginClient;
-    /** @var RequestFactory */
+    /** @var RequestFactory|RequestFactoryInterface */
     private $requestFactory;
     /** @var Plugin[] */
     private $plugins = [];
 
-    public function __construct(ClientInterface $httpClient = null, RequestFactory $requestFactory = null)
+    /**
+     * @param RequestFactory|RequestFactoryInterface|null $requestFactory
+     */
+    public function __construct(?ClientInterface $httpClient = null, $requestFactory = null)
     {
-        $this->httpClient = $httpClient ?: HttpClientDiscovery::find();
-        $this->requestFactory = $requestFactory ?: MessageFactoryDiscovery::find();
+        $requestFactory = $requestFactory ?: Psr17FactoryDiscovery::findRequestFactory();
```

Comment:
Should this be changed from `?:` to `??` so that a `TypeError` is thrown if `false` is passed in as the request factory?

In fact, most places with `?:` could be replaced now that PHP 7.2+ is required.

Reply:
Sure, using `??` here makes sense to keep enforcing the correct types.